### ISSUE TITLE
Eval bug: mistral4 empty output on Metal for prompts over ~300 tokens (clean GGUF, with and without flash attention)

### ISSUE THREAD

**Issue by jdrickman** [ggml-org/llama.cpp#25722]
### Name and Version

build ac4cdde (Metal), from Docker Model Runner's latest-metal backend image as of this posting (Docker Desktop 4.82.0, DMR v1.2.5). system_fingerprint: b1-ac4cdde

Updated: reproduced on DMR v1.2.6 / build 72874f5

### Operating systems

Mac

### GGML backends

Metal

### Hardware

Apple M5 Max, 128 GB unified memory

### Models

Mistral Small 4 119B (UD-IQ3_XXS): https://huggingface.co/unsloth/Mistral-Small-4-119B-2603-GGUF — single file, 42,797,027,008 bytes, downloaded 2026-07-15 (the post-regeneration upload, after the corruption reports in #20668 were addressed)

### Problem description & steps to reproduce

I ran into this while trying to use this model in DMR with Opencode, and spent a fair amount of time chasing my tail before enlisting the help of Claude to hammer through the possibilities. I automated the testing with Claude and then stepped through the process manually several times to ensure there were no hallucinations or prompt specific triggers, varying the context window size...which was probably a waste of time but I wanted to exhaust all options and try to capture as much useful information for triage as possible. 

Test matrix (OpenAI-compat server, same model/file; note: llama-completion is not available in this deployment — Docker Model Runner ships only the server binary, so all tests go through /v1/completions and /v1/chat/completions on the same engine):

| Test | -fa auto (default) | -fa off | -fa off + --fit off |
|---|---|---|---|
| /v1/completions, 6-token prompt, 30 max_tokens | empty | works, coherent text | works, coherent text |
| /v1/completions, 332-token prompt, 60 max_tokens | empty | empty (finish_reason: length) | empty (finish_reason: length) |
| /v1/chat/completions, trivial message (562-token templated prompt), 3000 max_tokens | empty | empty — 3000 tokens generated, content: "", no reasoning_content | not retested |

With --verbose, the long-prompt failure shows the model degenerating into emitting token id 31 for every generated token. Token 31 in this model's vocabulary is literally "<SPECIAL_31>" (type CONTROL, verified from the GGUF tokenizer metadata) — the same token reported as spam in #20668; the server strips control tokens from the response text, which is why the output appears empty rather than repetitive.

The long-prompt failure reproduces with flash attention off AND --fit off — a distinct defect from both #20710 (FA, fixed 2026-06-14, which postdates this build) and #20703 (--fit offload choice).

GGUF integrity check (gguf-py, np.isfinite + magnitude on all F32/F16/BF16 tensors): 181/181 clean, global max |x| = 6.78. blk.0.ffn_norm.weight — the tensor reported corrupted in #20668 — is normal in this file. Confirmed manually.

Repro:
1. llama-server -m Mistral-Small-4-119B-2603-UD-IQ3_XXS.gguf --ctx-size 40960 --jinja --flash-attn off --fit off
2. curl .../v1/completions -d '{"prompt": "The capital of France is", "max_tokens": 30}'  -> coherent continuation
3. curl .../v1/completions -d '{"prompt": "<repeat any sentence ~40x, ~330 tokens> Summarize in one sentence:", "max_tokens": 60}'  -> text: "", finish_reason: length, all 60 generated tokens are id 31
4. Any chat completion -> content: "" with full max_tokens consumed

Model loads clean: 41.9 GB resident, no memory pressure; prompt processing ~500 t/s, generation ~90 t/s. 

Related: #20668 (same symptoms, stale-closed), #20710, #20703, docker/model-runner#865 (Docker Model Runner is waiting on this upstream).

Tested on Docker Model Runner's pinned llama.cpp build (ac4cdde, 2026-06-10); rebuilding current master on this machine is not practical for me — if this is already fixed on master, a commit pointer would be appreciated.

### First Bad Commit

Unknown — symptoms reported since the mistral4 arch landed, per #20668 (2026-03-17).

### Relevant log output

$ curl -s http://localhost:12434/engines/v1/completions -d '{"model":"...","prompt":"The capital of France is","max_tokens":30}'
# -fa off --fit off, short prompt — OK:
{"choices":[{"text":" Paris. The capital of Germany is Berlin. The capital of Spain is Madrid. ...", ...}]}

$ curl -s http://localhost:12434/engines/v1/completions -d '{"model":"...","prompt":"<332-token prompt>","max_tokens":60}'
# -fa off --fit off, long prompt — empty:
{"choices":[{"text":"","finish_reason":"length"}],"usage":{"prompt_tokens":332,"completion_tokens":60}}

# --verbose during the long-prompt request: the model degenerates into emitting
# special token id 31 repeatedly (rendered as empty text) — matching the
# <SPECIAL_31> spam described in #20668:
token:    31 ''
token:    31 ''
token:    31 ''
token:    31 ''
token:    31 ''
... (repeats for all 60 generated tokens)
slot print_timing: id  3 | task 0 | graphs reused = 59
slot release: id 3 | task 0 | stop processing: n_tokens = 391, truncated = 0
Parsed message: {"role":"assistant","content":""}

$ curl -s http://localhost:12434/engines/v1/chat/completions -d '{"model":"...","messages":[{"role":"user","content":"Say exactly: hello world"}],"max_tokens":3000}'
# chat (fa off, fit default) — 3000 invisible tokens:
{"choices":[{"message":{"role":"assistant","content":""},"finish_reason":"length"}],"usage":{"prompt_tokens":562,"completion_tokens":3000}}

**Comment by jdrickman** [comment]
Reproduced on DMR v1.2.6 / build 72874f5

**Comment by twixr** [comment]
Same issue, also on M5 Max (64GB)

**Comment by kwakayama** [comment]
Reproduced on **Apple M4 Max / 128 GB**, macOS, Metal, with `lmstudio-community/Mistral-Small-4-119B-2603-GGUF` (Q4_K_M) — same empty output, `finish_reason: length`, no `reasoning_content`.

**The trigger is `n_ubatch`, not prompt length.** Holding the prompt fixed at 619 tokens and changing only `-ub`:

```
llama-server -m Mistral-Small-4-119B-2603-Q4_K_M-00001-of-00002.gguf \
  -c 2048 -b N -ub N --parallel 1
```

| `-ub` | prompt_tokens | output |
|---|---|---|
| 31 | 619 | `2+2 equals 4.` |
| 32 | 619 | *(empty)* |

Same prompt, same build, one integer apart. With `-ub 16` I get correct output at 75, 194 and 619 tokens — every length that returns empty at the default `-ub`.

This lines up with the analysis in #26223: `ne21_mm_id_min` is 32, so `-ub <= 31` keeps every micro-batch on the `mul_mv_id` path and the `mul_mm_id` kernel that narrows operands to f16 is never reached. I also verified #26223 red/green on this machine using its own test cases (details in that PR), so the two boundaries agree — one measured on a live model, one on the kernel test.

Two corrections to the framing here, both assumptions I made myself before measuring:

- **Prompt length is not the variable.** Long prompts fail only because they are chunked into micro-batches of `n_ubatch` (default 512). Cap `-ub` and length stops mattering.
- **Flash attention is not a factor.** `-ub 16` alone is sufficient, with FA left at its default. The `-fa off` results in the original report are explained by `llama-cli` and `llama-server` batching differently rather than by FA itself — I saw the same 6-token prompt succeed under `llama-cli` and return empty under `llama-server` with identical `-fa` settings.

**Workaround for anyone blocked:** `-ub 31` (or lower). Only prefill is slower; generation is unaffected, since decode is single-token and always stays on the safe path.

Scope: one model, one quant, Metal only. I did not inspect logits directly, so I can't independently confirm the NaN — only that the failure boundary sits exactly where #26223 predicts.
